### DISCUSSION TITLE
testing: "There is no current event loop in thread 'MainThread'."

### DISCUSSION THREAD

**Discussion by alexted** [fastapi/fastapi#6502] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code
I have several modules with unit-tests:
**test_sum.py**
```python
import pytest
from httpx import AsyncClient

pytestmark = pytest.mark.anyio


async def test_sum(app):
    async with AsyncClient(app=app, base_url='http://127.0.0.1:5000') as ac:
        response = await ac.post('/v1/sum', json={"x": 4, "y": 6})

    assert response.status_code == 200
    assert response.json() == {"sum": 10}
```

**test_items.py**
```python
from tests.expected_data import created_item
from tests.moks_data import items

pytestmark = pytest.mark.anyio


async def test_create_item(app):
    async with AsyncClient(app=app, base_url='http://test') as ac:
        response = await ac.post('/v1/items', json={
            "name": "TestName",
            "description": "Test description",
            "price": 100
        })

    assert response.status_code == 200
    assert response.json() == created_item


async def test_get_item(fill_db, app):
    async with AsyncClient(app=app, base_url='http://test') as ac:
        response = await ac.get('/v1/items/2')

    assert response.status_code == 200
    assert response.json() == items[1]
```

**conftest.py**
```python
import pytest
from alembic.command import downgrade, upgrade
from alembic.config import Config as AlembicConfig

from core.app.web import create_app
from core.data.postgres.engine import session_scope
from core.data.postgres.models import Base, Item
from tests.moks_data import items


@pytest.fixture
def app():
    app_instance = create_app()
    return app_instance

@pytest.fixture(scope='session')
def db_session():
    config = AlembicConfig('alembic.ini')
    config.attributes['configure_logger'] = False

    upgrade(config, 'head')

    yield 'on head'

    downgrade(config, 'base')


@pytest.fixture(autouse=True)
async def clear_data(db_session):
    yield 'I will clear tables for you'

    reference_value_tables = [f'{name}' for name in ('role',)]

    async with session_scope() as session:
        for name, table in Base.metadata.tables.items():
            if name not in reference_value_tables:
                await session.execute(table.delete())


@pytest.fixture
async def fill_db(event_loop):
    items_list = [Item(**item) for item in items]
    async with session_scope() as session:
        session.add_all(items_list)
        await session.flush()
```

### Description
I'm using:
python 3.10.1, 
fastapi 0.72.0, 
pytest 6.2.5, 
pytest-trio 0.7.0, 
sqlalchemy 1.4.31, 
alembic 1.7.5,
asyncpg 0.25.0,
postgresql 14.1.**
Now, i have issue when i try to execute my tests.

When i run my tests - tests in test_sum is successfully ended but tests in test_items is failed with error:
`There is no current event loop in thread 'MainThread'.`

Please tell me what I'm doing wrong and how to solve this problem?

### Operating System

Linux

### Operating System Details

manjaro linux 21.2.1

### FastAPI Version

0.72.0

### Python Version

3.10.1

### Additional Context

tests work correctly with **fastapi <= 0.68**

**Comment by jgould22** [comment]
Your formatting makes things hard to read.

Is there a reason you are avoiding the use of the decorator that starts a loop for you?

**Comment by alexted** [comment]
@jgould22 ty, fixed!
Which decorator are we talking about? I just write unit tests as it says in the official fastapi documentation. I don't remember it talking about anything like that.
Isn't this handled by the under-the-hood framework?

**Comment by jgould22** [comment]
Specifically I meant `@pytest.mark.anyio`

You can see it in use here https://fastapi.tiangolo.com/advanced/async-tests/

However now I see you are using the `pytestmark = pytest.mark.anyio` method

Anyway `test_sum.py` probably passes because you are not using the async fixtures in `conftest.py`

In `test_items.py` you are using `fill_db` which is in `conftest.py` but you are not starting a loop in conftest (no `pytestmark = pytest.mark.anyio` in there)

Maybe try adding that to conf_test or wrapping the `fill_db` and `clear_data` in `@pytest.mark.anyio` so that fixture has a loop to use.

**Comment by mateusz-chrzastek** [comment]
Hey @alexted, have you ever resolved this?

**Comment by alexted** [comment]
@mateusz-chrzastek After a few patches, it seems to have fixed itself.

**Comment by mateusz-chrzastek** [comment]
I encountered this when writing my tests as asynchronous tests (`async def test_xyz`) with asyncpg and pytest-asyncio.

I found an alternate solution by opting for non-async `def test_xyz` tests and removing any class wrapper! (Which is what broke regular tests for me). While this doesn't resolve being able to run async code in tests, I was able to test my endpoints this way.

**Comment by alexted** [comment]
this material helped me a lot: https://rogulski.it/blog/sqlalchemy-14-async-orm-with-fastapi/
